Support for Python 3.10 tests

Nose library for unit tests is deprecated and does not support Python 3.10.

```
Run nosetests -vv --with-coverage --cover-package=libcloudforensics
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.plugins.cover: INFO: Coverage report will include only packages: ['libcloudforensics']
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.11/x[6](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:7)4/bin/nosetests", line 8, in <module>
    sys.exit(run_exit())
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/core.py", line 118, in __init__
    unittest.TestProgram.__init__(
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/unittest/main.py", line 100, in __init__
    self.parseArgs(argv)
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/core.py", line 1[7](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:8)9, in parseArgs
    self.createTests()
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/core.py", line 193, in createTests
    self.test = self.testLoader.loadTestsFromNames(self.testNames)
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/loader.py", line 4[8](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:9)1, in loadTestsFromNames
    return unittest.TestLoader.loadTestsFromNames(self, names, module)
  File "/opt/hostedtoolcache/Python/3.[10](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:11).[11](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:12)/x64/lib/python3.10/unittest/loader.py", line 2[20](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:21), in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/unittest/loader.py", line [22](https://github.com/jleaniz/cloud-forensics-utils/actions/runs/4928021894/jobs/8805836995#step:5:23)0, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/loader.py", line 454, in loadTestsFromName
    return LazySuite(
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/suite.py", line 53, in __init__
    super(LazySuite, self).__init__()
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/unittest/suite.py", line 22, in __init__
    self._tests = []
  File "/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/nose/suite.py", line 106, in _set_tests
    if isinstance(tests, collections.Callable) and not is_suite:
AttributeError: module 'collections' has no attribute 'Callable'
Error: Process completed with exit code 1.
```

nose usage should be removed in favor of unittest or pytest -- which would allow running the tests in Python 3.10

--- a/.github/workflows/install_test.yml
+++ b/.github/workflows/install_test.yml
@@ -27,4 +27,4 @@ jobs:
         pip install -r requirements-dev.txt
     - name: Test with nose
       run: |
-        nosetests -vv --with-coverage --cover-package=libcloudforensics
+        python -m coverage run -m unittest

--- a/.github/workflows/install_test.yml
+++ b/.github/workflows/install_test.yml
@@ -12,7 +12,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest]
-        python-version: [3.7, 3.8, 3.9]
+        python-version: ['3.7', '3.8', '3.9', '3.10', '3.11']
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/install_test.yml
+++ b/.github/workflows/install_test.yml
@@ -25,6 +25,6 @@ jobs:
         python -m pip install --upgrade pip
         pip install -r requirements.txt
         pip install -r requirements-dev.txt
-    - name: Test with nose
+    - name: Test with unittest
       run: |
         python -m coverage run -m unittest

--- a/.github/workflows/install_test_coverage.yml
+++ b/.github/workflows/install_test_coverage.yml
@@ -8,24 +8,28 @@ jobs:
   build:
 
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        python-version: [3.7, 3.8, 3.9, 3.10, 3.11]
 
     steps:
     - uses: actions/checkout@v2
-    - name: Set up Python 3.7
+    - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v1
       with:
-        python-version: 3.7
+        python-version: ${{ matrix.python-version }}
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
         pip install -r requirements.txt
         pip install -r requirements-dev.txt
-    - name: Test with nose
+    - name: Test
       run: |
-        nosetests -vv --with-coverage --cover-xml --cover-package=libcloudforensics
+        python -m coverage run -m unittest
     - name: Upload coverage to Codecov
       uses: codecov/codecov-action@v1
       with:
         file: ./coverage.xml
-        flags: nosetests
+        flags: unittests
         name: codecov-umbrella

--- a/.github/workflows/install_test_coverage.yml
+++ b/.github/workflows/install_test_coverage.yml
@@ -11,7 +11,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest]
-        python-version: [3.7, 3.8, 3.9, 3.10, 3.11]
+        python-version: ['3.7', '3.8', '3.9', '3.10', '3.11']
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/install_test_setup_py.yml
+++ b/.github/workflows/install_test_setup_py.yml
@@ -25,7 +25,7 @@ jobs:
       run: pip install wheel
     - name: Install dependencies through setup.py
       run: python setup.py install
-    - name: Test through setup.py
+    - name: Test
       run: |
         pip install mock
         python -m unittest

--- a/.github/workflows/install_test_setup_py.yml
+++ b/.github/workflows/install_test_setup_py.yml
@@ -11,7 +11,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest]
-        python-version: [3.7, 3.8, 3.9, 3.10, 3.11]
+        python-version: ['3.7', '3.8', '3.9', '3.10', '3.11']
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/install_test_setup_py.yml
+++ b/.github/workflows/install_test_setup_py.yml
@@ -11,7 +11,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest]
-        python-version: [3.7, 3.8, 3.9]
+        python-version: [3.7, 3.8, 3.9, 3.10, 3.11]
 
     steps:
     - uses: actions/checkout@v2
@@ -27,5 +27,5 @@ jobs:
       run: python setup.py install
     - name: Test through setup.py
       run: |
-        pip install nose mock
-        nosetests -vv
+        pip install mock
+        python -m unittest

--- a/.pylintrc
+++ b/.pylintrc
@@ -56,17 +56,11 @@ disable=
     consider-using-f-string,
     deprecated-pragma,
     duplicate-code,
-    eq-without-hash,
     file-ignored,
     fixme,
     locally-disabled,
-    locally-enabled,
     logging-format-interpolation,
-    metaclass-assignment,
     missing-param-doc,
-    no-absolute-import,
-    no-self-use,
-    parameter-unpacking,
     raw-checker-failed,
     suppressed-message,
     too-few-public-methods,
@@ -210,17 +204,9 @@ logging-modules=logging
 
 [BASIC]
 
-# Naming hint for argument names
-# argument-name-hint=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
-argument-name-hint=(([a-z][a-z0-9_]*)|(_[a-z0-9_]*))$
-
 # Regular expression matching correct argument names
 # argument-rgx=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
 argument-rgx=(([a-z][a-z0-9_]*)|(_[a-z0-9_]*))$
-
-# Naming hint for attribute names
-# attr-name-hint=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
-attr-name-hint=(([a-z][a-z0-9_]*)|(_[a-z0-9_]*))$
 
 # Regular expression matching correct attribute names
 # attr-rgx=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
@@ -229,23 +215,12 @@ attr-rgx=(([a-z][a-z0-9_]*)|(_[a-z0-9_]*))$
 # Bad variable names which should always be refused, separated by a comma
 bad-names=foo,bar,baz,toto,tutu,tata
 
-# Naming hint for class attribute names
-# class-attribute-name-hint=([A-Za-z_][A-Za-z0-9_]{2,30}|(__.*__))$
-class-attribute-name-hint=([A-Za-z_][A-Za-z0-9_]*|(__.*__))$
-
 # Regular expression matching correct class attribute names
 # class-attribute-rgx=([A-Za-z_][A-Za-z0-9_]{2,30}|(__.*__))$
 class-attribute-rgx=([A-Za-z_][A-Za-z0-9_]*|(__.*__))$
 
-# Naming hint for class names
-class-name-hint=[A-Z_][a-zA-Z0-9]+$
-
 # Regular expression matching correct class names
 class-rgx=[A-Z_][a-zA-Z0-9]+$
-
-# Naming hint for constant names
-# const-name-hint=(([A-Z_][A-Z0-9_]*)|(__.*__))$
-const-name-hint=(([a-zA-Z_][a-zA-Z0-9_]*)|(__.*__))$
 
 # Regular expression matching correct constant names
 # const-rgx=(([A-Z_][A-Z0-9_]*)|(__.*__))$
@@ -254,10 +229,6 @@ const-rgx=(([a-zA-Z_][a-zA-Z0-9_]*)|(__.*__))$
 # Minimum line length for functions/classes that require docstrings, shorter
 # ones are exempt.
 docstring-min-length=-1
-
-# Naming hint for function names
-# function-name-hint=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
-function-name-hint=[A-Z_][a-zA-Z0-9_]*$
 
 # Regular expression matching correct function names
 # function-rgx=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
@@ -269,22 +240,13 @@ good-names=i,j,k,ex,Run,_
 # Include a hint for the correct naming format with invalid-name
 include-naming-hint=no
 
-# Naming hint for inline iteration names
-inlinevar-name-hint=[A-Za-z_][A-Za-z0-9_]*$
-
 # Regular expression matching correct inline iteration names
 inlinevar-rgx=[A-Za-z_][A-Za-z0-9_]*$
 
-# Naming hint for method names
-# method-name-hint=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
-method-name-hint=(test|[A-Z_])[a-zA-Z0-9_]*$
 
 # Regular expression matching correct method names
 # method-rgx=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
 method-rgx=(test|[A-Z_])[a-zA-Z0-9_]*$
-
-# Naming hint for module names
-module-name-hint=(([a-z_][a-z0-9_]*)|([A-Z][a-zA-Z0-9]+))$
 
 # Regular expression matching correct module names
 module-rgx=(([a-z_][a-z0-9_]*)|([A-Z][a-zA-Z0-9]+))$
@@ -300,10 +262,6 @@ no-docstring-rgx=^_
 # List of decorators that produce properties, such as abc.abstractproperty. Add
 # to this list to register other decorators that produce valid properties.
 property-classes=abc.abstractproperty
-
-# Naming hint for variable names
-# variable-name-hint=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
-variable-name-hint=(([a-z][a-z0-9_]*)|(_[a-z0-9_]*))$
 
 # Regular expression matching correct variable names
 # variable-rgx=(([a-z][a-z0-9_]{2,30})|(_[a-z0-9_]*))$
@@ -338,12 +296,6 @@ max-line-length=80
 
 # Maximum number of lines in a module
 max-module-lines=1000
-
-# List of optional constructs for which whitespace checking is disabled. `dict-
-# separator` is used to allow tabulation in dicts, etc.: {1  : 1,\n222: 2}.
-# `trailing-comma` allows a space between comma and closing bracket: (a, ).
-# `empty-line` allows space-only lines.
-no-space-check=trailing-comma,dict-separator
 
 # Allow the body of a class to be on the same line as the declaration if body
 # contains single statement.
@@ -472,4 +424,4 @@ known-third-party=enchant
 
 # Exceptions that will emit a warning when being caught. Defaults to
 # "Exception"
-overgeneral-exceptions=Exception
+overgeneral-exceptions=builtins.Exception

--- a/libcloudforensics/providers/aws/internal/iam.py
+++ b/libcloudforensics/providers/aws/internal/iam.py
@@ -17,7 +17,7 @@
 import datetime
 import json
 import os
-from typing import TYPE_CHECKING, Optional, Tuple
+from typing import TYPE_CHECKING, Tuple
 from libcloudforensics import errors
 from libcloudforensics import logging_utils
 from libcloudforensics.providers.aws.internal import common

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -2,7 +2,6 @@
 coverage
 mock
 mypy
-nose
 pylint
 sphinx
 sphinx-autodoc-typehints

--- a/requirements.txt
+++ b/requirements.txt
@@ -8,8 +8,8 @@ azure-mgmt-network
 azure-mgmt-resource
 azure-mgmt-storage
 azure-storage-blob
-boto3
-botocore
+boto3>=1.26.135
+botocore>=1.29.135
 google-api-core
 google-api-python-client
 # google-auth: We don't directly care about this but 2.0.0dev0 somehow gets

--- a/requirements.txt
+++ b/requirements.txt
@@ -20,10 +20,11 @@ msrest
 msrestazure
 netaddr
 pycryptodome
-PyJWT<2  # Bad overlap in azure dependencies
+PyJWT
 sshpubkeys
 requests
 kubernetes
 # pyopenssl: some environments include old versions which can cause issues with
 # packages installed by boto3
 pyopenssl
+urllib3<1.27,>=1.21.1

--- a/requirements.txt
+++ b/requirements.txt
@@ -14,7 +14,7 @@ google-api-core
 google-api-python-client
 # google-auth: We don't directly care about this but 2.0.0dev0 somehow gets
 # installed despite the pinning in google-api-core
-google-auth<2.0.0.dev0
+# google-auth<2.0.0.dev0
 httplib2
 msrest
 msrestazure

--- a/tests/providers/azure/internal/test_monitoring.py
+++ b/tests/providers/azure/internal/test_monitoring.py
@@ -26,7 +26,7 @@ class AZMonitoringTest(unittest.TestCase):
   # pylint: disable=line-too-long
 
   @typing.no_type_check
-  @mock.patch('azure.mgmt.monitor.v2018_01_01.operations._metric_definitions_operations.MetricDefinitionsOperations.list')
+  @mock.patch('azure.mgmt.monitor.v2021_05_01.operations._metric_definitions_operations.MetricDefinitionsOperations.list')
   def testListAvailableMetricsForResource(self, mock_list_metrics):
     """Test that metrics are correctly listed."""
     mock_list_metrics.return_value = azure_mocks.MOCK_LIST_METRICS
@@ -36,7 +36,7 @@ class AZMonitoringTest(unittest.TestCase):
     self.assertIn('fake-metric', metrics)
 
   @typing.no_type_check
-  @mock.patch('azure.mgmt.monitor.v2018_01_01.operations._metrics_operations.MetricsOperations.list')
+  @mock.patch('azure.mgmt.monitor.v2021_05_01.operations._metrics_operations.MetricsOperations.list')
   def testGetMetricsForResource(self, mock_list_metrics_operations):
     """Test that metrics values are correctly retrieved."""
     mock_list_metrics_operations.return_value = azure_mocks.MOCK_METRIC_OPERATION

--- a/tests/providers/azure/internal/test_storage.py
+++ b/tests/providers/azure/internal/test_storage.py
@@ -26,8 +26,8 @@ class AZStorageTest(unittest.TestCase):
   """Test Azure storage class."""
   # pylint: disable=line-too-long
 
-  @mock.patch('azure.mgmt.storage.v2022_05_01.operations._storage_accounts_operations.StorageAccountsOperations.list_keys')
-  @mock.patch('azure.mgmt.storage.v2022_05_01.operations._storage_accounts_operations.StorageAccountsOperations.begin_create')
+  @mock.patch('azure.mgmt.storage.v2022_09_01.operations._storage_accounts_operations.StorageAccountsOperations.list_keys')
+  @mock.patch('azure.mgmt.storage.v2022_09_01.operations._storage_accounts_operations.StorageAccountsOperations.begin_create')
   @typing.no_type_check
   def testCreateStorageAccount(self, mock_create, mock_list_keys):
     """Test that a storage account is created and its information retrieved"""

--- a/tools/aws_cli.py
+++ b/tools/aws_cli.py
@@ -27,8 +27,6 @@ from datetime import datetime
 from typing import TYPE_CHECKING
 
 from libcloudforensics.providers.aws.internal import account
-from libcloudforensics.providers.aws.internal import ec2
-from libcloudforensics.providers.aws.internal import iam
 from libcloudforensics.providers.aws.internal import log as aws_log
 from libcloudforensics.providers.aws import forensics
 from libcloudforensics import logging_utils

